File: s08_context_compact/permission.py
from anthropic.types import ContentBlock, ToolUseBlock
from loguru import logger
from rich import print
from rich.prompt import Confirm
from typing_extensions import Literal

from .constants import WORKDIR

DENY_LIST = [
    "rm -rf / --",
    "sudo",
    "shutdown",
    "reboot",
    "mkfs",
    "dd if=",
    "> /dev/",
]


# Gate 1: Deny List
def check_deny_list(command: str) -> str | None:
    for pattern in DENY_LIST:
        if pattern in command:
            return f"Blocked: '{pattern}' is on the deny list"
    return None


# 权限规则列表，每条规则包含：适用工具、检查函数（lambda）、拦截提示
PERMISSION_RULES = [
    {
        "tools": ["write_file", "edit_file"],
        # 将参数中的 path 拼接到工作目录后，resolve() 解析为绝对路径，
        # 再用 is_relative_to() 判断是否仍在 WORKDIR 内；
        # 若不在（not ...），说明试图写入工作区之外，返回 True 触发拦截
        "check": lambda args: (
            not (WORKDIR / args.get("path", "")).resolve().is_relative_to(WORKDIR)
        ),
        "message": "Writing outside workspace",
    },
    {
        "tools": ["bash"],
        # 从参数中取出 command 字符串，用 any() 逐一检查是否包含危险关键词；
        # 只要命中任意一个（"rm "、"> /etc/"、"chmod 777"），返回 True 触发拦截
        "check": lambda args: any(
            kw in args.get("command", "") for kw in ["rm ", "> /etc/", "chmod 777"]
        ),
        "message": "Potentially destructive command",
    },
]


def check_rules(tool_name: str, args: dict) -> str | None:
    """Gate 2: 遍历 PERMISSION_RULES，若当前工具命中某条规则且 check 返回 True，则返回拦截信息"""
    for rule in PERMISSION_RULES:
        if tool_name in rule["tools"] and rule["check"](args):
            return rule["message"]
    return None


def ask_user(tool_name: str, args: dict, reason: str) -> Literal["allow", "deny"]:
    """
    Gate 3: User approval — wait for confirmation after rule match
    """
    logger.warning("⚠  {}: {}({})", reason, tool_name, args)
    allowed = Confirm.ask("   Allow?", default=False)
    return "allow" if allowed else "deny"


# Pipeline: all three gates chained
def permission_hook(block: ToolUseBlock) -> str | None:
    # 1
    if block.name == "bash":
        reason = check_deny_list(block.input.get("command", "").__str__())
        if reason:
            logger.error("Permission denied: {}", reason)
            return "Permission denied by deny list."
    # 2
    reason = check_rules(block.name, block.input)
    if reason is not None:
        # 3
        decision = ask_user(block.name, block.input, reason)
        if decision == "deny":
            return "Permission denied by user."
    return None

File: permission.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/anthropics/anthropic-sdk-go"
)

// ── 拒绝列表（Gate 1）──────────────────────────────────

var denyList = []string{
	"rm -rf / --",
	"sudo",
	"shutdown",
	"reboot",
	"mkfs",
	"dd if=",
	"> /dev/",
}

// checkDenyList 检查命令是否命中绝对禁止列表
func checkDenyList(command string) *string {
	for _, pattern := range denyList {
		if strings.Contains(command, pattern) {
			msg := fmt.Sprintf("Blocked: '%s' is on the deny list", pattern)
			return &msg
		}
	}
	return nil
}

// ── 规则检查（Gate 2）──────────────────────────────────

type permissionRule struct {
	tools   []string
	check   func(input json.RawMessage) bool
	message string
}

var permissionRules = []permissionRule{
	{
		tools: []string{"write_file", "edit_file"},
		check: func(raw json.RawMessage) bool {
			var input struct {
				Path string `json:"path"`
			}
			json.Unmarshal(raw, &input)
			_, err := safePath(input.Path)
			return err != nil // 路径逃逸 → 触发拦截
		},
		message: "Writing outside workspace",
	},
	{
		tools: []string{"bash"},
		check: func(raw json.RawMessage) bool {
			var input struct {
				Command string `json:"command"`
			}
			json.Unmarshal(raw, &input)
			for _, kw := range []string{"rm ", "> /etc/", "chmod 777"} {
				if strings.Contains(input.Command, kw) {
					return true
				}
			}
			return false
		},
		message: "Potentially destructive command",
	},
}

// checkRules 遍历权限规则，返回命中的拦截信息
func checkRules(toolName string, input json.RawMessage) *string {
	for _, rule := range permissionRules {
		for _, t := range rule.tools {
			if t == toolName && rule.check(input) {
				return &rule.message
			}
		}
	}
	return nil
}

// ── 用户确认（Gate 3）──────────────────────────────────

// askUser 在规则命中后询问用户是否允许
func askUser(toolName string, input json.RawMessage, reason string) bool {
	fmt.Printf("%s ⚠  %s: %s(%s)\n", colorYellow("WARNING"), reason, toolName, truncate(string(input), 60))
	fmt.Print("   Allow? [y/N]: ")

	scanner := bufio.NewScanner(os.Stdin)
	if scanner.Scan() {
		answer := strings.TrimSpace(strings.ToLower(scanner.Text()))
		return answer == "y" || answer == "yes"
	}
	return false
}

// ── 三层门控串联 ──────────────────────────────────────

// permissionHook 作为 PreToolUse 钩子，串联三层权限检查
// 返回 nil 表示放行，返回 *string 表示拒绝（内容作为 tool_result 返回给模型）
func permissionHook(args ...any) *string {
	block, ok := args[0].(anthropic.ToolUseBlock)
	if !ok {
		return nil
	}

	rawInput, _ := json.Marshal(block.Input)

	// Gate 1: 拒绝列表
	if block.Name == "bash" {
		var input struct {
			Command string `json:"command"`
		}
		json.Unmarshal(rawInput, &input)
		if reason := checkDenyList(input.Command); reason != nil {
			msg := "Permission denied by deny list."
			fmt.Println(colorRed("ERROR"), *reason)
			return &msg
		}
	}

	// Gate 2: 规则检查
	if reason := checkRules(block.Name, rawInput); reason != nil {
		// Gate 3: 用户确认
		if !askUser(block.Name, rawInput, *reason) {
			msg := "Permission denied by user."
			return &msg
		}
	}

	return nil
}

// truncate 截断字符串用于预览
func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
